packages/jazz-tools/src/coValues/coMap.ts

```diff
@@ -564,6 +563,80 @@ export class CoMap extends CoValueBase implements CoValue {
     return cojsonInternals.idforHeader(header, crypto) as ID<M>;
   }
 
+  /**
+   * Given some data, updates an existing CoMap or initialises a new one if none exists.
+   *
+   * @example
+   * ```ts
+   * const activeEvent = await Event.upsertUnique(
+   *   sourceData.identifier,
+   *   workspace.id,
+   *   {
+   *     title: sourceData.title,
+   *     identifier: sourceData.identifier,
+   *     external_id: sourceData._id,
+   *   },
+   *   workspace
+   * );
+   * ```
+   *
+   * @param unique The data that uniquely identifies the CoMap.
+   * @param ownerID The ID of the owner of the CoMap.
+   * @param init The data that the CoMap should contain.
+   * @param creationOptions The creation options for the CoMap.
+   * @param loadOptions The options for loading the CoMap, if it exists (note: `skipRetry` defaults to `true`).
+   * @returns Either an existing & modified CoMap, or a new initialised CoMap if none exists.
+   * @category Subscription & Loading
+   */
+  static async upsertUnique<
+    M extends CoMap,
+    const R extends RefsToResolve<M> = true,
+  >(
+    this: CoValueClass<M>,
+    unique: CoValueUniqueness["uniqueness"],
+    ownerID: ID<Account> | ID<Group>,
+    init: Simplify<CoMapInit<M>>,
+    creationOptions?:
+      | {
+          owner: Account | Group;
+        }
+      | Account
+      | Group,
+    loadOptions?: {
+      resolve?: RefsToResolveStrict<M, R>;
+      loadAs?: Account | AnonymousJazzAgent;
+      skipRetry?: boolean;
+    },
+  ): Promise<M> {
+    let mapId = CoMap.findUnique(unique, ownerID);
+    let map: Resolved<M, true> | null = await loadCoValueWithoutMe(
+      this,
+      mapId,
+      { ...loadOptions, skipRetry: loadOptions?.skipRetry ?? true },
+    );
+    if (!map) {
+      const instance = new this();
+      const createOptions =
+        creationOptions && !("_type" in creationOptions)
+          ? { ...creationOptions, unique: unique }
+          : creationOptions;
+      const { owner, uniqueness } = parseCoValueCreateOptions(createOptions);
+      const raw = instance.rawFromInit(init, owner, uniqueness);
+      Object.defineProperties(instance, {
+        id: {
+          value: raw.id,
+          enumerable: false,
+        },
+        _raw: { value: raw, enumerable: false },
+      });
+      map = instance;
+    } else {
+      map.applyDiff(init as Partial<CoMapInit<M>>);
+    }
+
+    return map as M;
```

Comment:
We need to call `loadCoValueWithoutMe` again to ensure that the applied update respects the `resolve` param.